onedal/decomposition/pca.py

```diff
@@ -0,0 +1,101 @@
+#===============================================================================
+# Copyright 2023 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#===============================================================================
+from onedal import _backend
+
+from ..common._policy import _get_policy
+from ..datatypes._data_conversion import from_table, to_table
+from daal4py.sklearn._utils import sklearn_check_version
+
+import numpy as np
+
+
+class PCA():
+    def __init__(
+        self,
+        n_components=None,
+        is_deterministic=True,
+        method='precomputed',
+        copy=True
+    ):
+        self.n_components = n_components
+        self.copy = copy
+        self.method = method
+        self.is_deterministic = is_deterministic
+
+    def get_onedal_params(self, data):
+        return {
+            'fptype':
+                'float' if data.dtype is np.dtype('float32') else 'double',
+            'method': self.method,
+            'n_components': self.n_components,
+            'is_deterministic': self.is_deterministic
+        }
+
+    def fit(self, X, y, queue):
+        n_samples, n_features = X.shape
+        n_sf_min = min(n_samples, n_features)
+
+        policy = _get_policy(queue, X, y)
+        params = self.get_onedal_params(X)
+        cov_result = _backend.covariance.compute(
+            policy,
+            {'fptype': params['fptype'], 'method': 'dense'},
+            to_table(X)
+        )
+        covariance_matrix = from_table(cov_result.cov_matrix)
+        result = _backend.decomposition.dim_reduction.train(
+            policy,
+            params,
+            to_table(covariance_matrix)
+        )
+
+        self.n_components_ = self.n_components
+        self.variances_ = from_table(result.variances)
+        self.components_ = from_table(result.eigenvectors)
+        self.explained_variance_ = from_table(result.eigenvalues)
+        tot_var = covariance_matrix.trace()
+        self.explained_variance_ratio_ = self.explained_variance_ / tot_var
+        self.singular_values_ = np.sqrt(
+            (n_samples - 1) * self.explained_variance_
+        )
+
+        if sklearn_check_version("1.2"):
+            self.n_features_in_ = n_features
+        else:
+            self.n_features_ = n_features
```

Comment:
`sklearn/tests/test_common.py::test_check_n_features_in_after_fitting[PCA()]` failed on sklearn>0.24 because of `n_features_in_` absence. I guess assignment should be added for earlier versions than 1.2